### PR TITLE
Add unbond button

### PR DESCRIPTION
**Pull request by Tbaut**
Addresses the `unbond` part in https://github.com/polkadot-js/apps/issues/859

This PR also addresses 
- a little quirk that the "Bond Additional" button was still visible although there wasn't anything possible to bond.
- removes an unused `validators: Array<string>` type.
- ordering imports alphabetically 

I'll prepare a separate PR to address the `withdrawUnbonded` part of https://github.com/polkadot-js/apps/issues/859 to make things easier to review.

### CODE REVIEW

**Comment by Tbaut** [comment]
Thanks, looking forward for the review, I wonder if there is a nice was to group the `staking.ledger` calls added here: https://github.com/polkadot-js/apps/pull/1057/files#diff-11cd827cd9b9d5e0d8503e090149dcacR464 I couldn't figure this out.

**Comment by jacogr** [comment]
Not really, unless you return both in a single call. 

```js
['query.staking.ledger', {
      paramName: 'controllerId',
      propName: 'stashBonded',
      transform: (ledger: Option<StakingLedger>) =>
        ledger.unwrapOr({ active: null, total: null })
    }],
```

It is generally not too bad since the calls are cached on the API layer, so at the very least it doesn't do multiples.

File: packages/app-staking/src/Account/index.tsx [hunk 1]
```diff
@@ -118,6 +121,19 @@ class Account extends React.PureComponent<Props, State> {
...
   }
 
+  private renderUnbond () {
+    const { controllerId } = this.props;
+    const { isUnbondOpen } = this.state;
+
+    return (
+        <Unbond
+          controllerId={controllerId}
+          isOpen={isUnbondOpen}
```
Comment (Stefie) [inline packages/app-staking/src/Account/index.tsx]: There's  no need to pass`isOpen` down to the component. You can just conditionally render  `<Unbond>` it here in the index.tsx (see comment above)

File: packages/app-staking/src/Account/index.tsx [hunk 2]
```diff
@@ -69,6 +71,7 @@ class Account extends React.PureComponent<Props, State> {
         {this.renderBondExtra()}
         {this.renderNominating()}
         {this.renderSessionKey()}
+        {this.renderUnbond()}
```
Comment (Stefie) [inline packages/app-staking/src/Account/index.tsx]: Rather do `{this.state.isUnbondOpen  && this.renderUnbond()}` (or deconstruct this.state.isUnbondOpen before calling it).
 
There's no need to mount the `Unbond ` component, pass down the `isUnbondOpen` prop and then let the component return null if `isUnbondOpen` is false. It's better to just not call `this.renderUnbond()`  which mounts `<Unbond>`  at all if `isUnbondOpen` is false.

Same is true for all the other calls, so If you're up for refactoring tha in a second PR, feel free :).

File: packages/app-staking/src/Account/Unbond.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  state: State = {};
+
+  render () {
+    const { controllerId, isOpen, onClose, t } = this.props;
+    const { maxUnbond } = this.state;
+    const canSubmit = !!maxUnbond && maxUnbond.gtn(0);
+
+    if (!isOpen) {
+      return null;
+    }
```
Comment (Stefie) [inline packages/app-staking/src/Account/Unbond.tsx]: See my comments regarding `isOpen` on the index.js. It's cleaner to conditionally render this component in the index.js instead of retuning `null` here.

File: packages/app-staking/src/Account/Unbond.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import BN from 'bn.js';
+import React from 'react';
+import { AccountId } from '@polkadot/types';
+import { Button, InputAddress, InputBalance, Modal, TxButton } from '@polkadot/ui-app';
+
+import translate from '../translate';
+
+type Props = I18nProps & {
+  controllerId?: AccountId | null,
+  isOpen: boolean,
```
Comment (Stefie) [inline packages/app-staking/src/Account/Unbond.tsx]: Remove prop (see comment on index.js)

**Comment by Tbaut** [comment]
Thanks for the reviews, although for this PR I didn't originally want to refactor the isOpen things, it's small enough that I'm confident not to break everything :)

File: packages/app-staking/src/Account/index.tsx [hunk 5]
```diff
@@ -294,13 +299,14 @@ class Account extends React.PureComponent<Props, State> {
     }
 
     return (
+      isNominateOpen ?
```
Comment (jacogr) [inline packages/app-staking/src/Account/index.tsx]: There _are_ actually very good reasons why some of these are done this way. 

In the cases where we do queries and display the info to the user, we want to fire them in the background, so there is no actual delay for the display of the info. It has the additional benefit that the subscriptions are not re-created each time as well.

(The latter probably only until we get multi support inside the API)

File: packages/app-staking/src/Account/index.tsx [hunk 6]
```diff
@@ -58,18 +59,20 @@ class Account extends React.PureComponent<Props, State> {
...
     }
 
     return (
       <article className='staking--Account'>
-        {this.renderBond()}
-        {this.renderBondExtra()}
-        {this.renderNominating()}
-        {this.renderSessionKey()}
-        {this.renderValidating()}
+        {isBondOpen && this.renderBond()}
```
Comment (jacogr) [inline packages/app-staking/src/Account/index.tsx]: Re-copying the comment here.

There are actually very good reasons why some of these are done this way with the passing through of `isOpen`.

In the cases where we do queries and display the info to the user, we want to fire them in the background, so there is no actual delay for the display of the info. It has the additional benefit that the subscriptions are not re-created each time as well.

(The latter probably only until we get multi support inside the API)

... and all obviously not obvious as to why

Comment (Tbaut) [reply packages/app-staking/src/Account/index.tsx]: Damn I answered on the original comment but it didn't get through.

Thanks for the info, very interesting. Before I revert it all you say "why **some** of these are done this way". Should I keep this "new" behaviour (without passing the isOpen) for some components?

**Comment by jacogr** [comment]
I would honestly just pass through isOpen on all

File: packages/app-staking/src/Account/index.tsx (hunk 6 above)
Comment (Stefie) [reply packages/app-staking/src/Account/index.tsx]: Hmm, but none of them are actually doing any queries (as in API queries) in the background, or am I missing something?
```
{this.renderBond()} -> src/Account/Bond.tsx
{this.renderBondExtra()} -> src/Account/BondExtra.tsx
{this.renderNominating()} -> src/Account/Nominating.tsx
{this.renderSessionKey()} -> src/Account/SessionKey.tsx
{this.renderValidating()} -> src/Account/Validating.tsx
```
The only component doing an API call in ./Accounts is the `ValidateController.tsx` which is used in `Account/Bond.tsx` and called from within `renderBond()` .

The way it is implemented we're doing an an api all in `ValidateController` for every account listed under the  'Account Actions' tab, even if the user never opens the Bond overlay. That would make one unneeded call for every account as opposed to maybe re-creating a subsription for the same account in case the user is opening the same overlay twice (or more)...

**Comment by Tbaut** [comment]
We're back with the original logic and a comment that is hopefully clear enough.

**Comment by Stefie** [comment]
@jacogr  Did you read my comment above? We'Re not using `withCalls()` in any of the components where i suggested the change.

```
{this.renderBond()} -> src/Account/Bond.tsx
{this.renderBondExtra()} -> src/Account/BondExtra.tsx
{this.renderNominating()} -> src/Account/Nominating.tsx
{this.renderSessionKey()} -> src/Account/SessionKey.tsx
{this.renderValidating()} -> src/Account/Validating.tsx
```

**Comment by jacogr** [comment]
I always prefer consistency over guessing here. Additionally, we are only about 50% through as it stands - we had to get staking updated fast. So for the modals which I can think of offhand - 

Bond - should have the free and bonded balances as well, and only allow access to what is available. Here we also need to retrieve any previously set preferences (not sure but I believe I dropped that at the last moment)

Bond additional - Same in regards to funds. (Although, thankfully for the time being the max is also adjusted on the node)

Session - should additionally check keys for all accounts, so that you don’t assign. 

The 2 bigger ones for validation and nomination are a bit more problematic. Here we need more info on a per account basis, and atm the current approach is not quite optimal for that. 

(Not sure if I missed some atm)

**Comment by Stefie** [comment]
Hmm... still not really convinced by that approach. 
If we'll need the infos you listed for all accounts in the future, we should do the api calls in the index.tsx and just pass the results down as props to the components instead of mounting all possible modal components without displaying them just to do the api calls. That still feels like a unnecessary detour to me...

But that discussion hasn't much to do with @Tbaut's PR anymore (sorry for delaying your merge :))

**Comment by jacogr** [comment]
In all honesty, trying to do too much atm (like changing the modals) is wasted work. All the modals probably need to go the way of the dodo and we need to have a unified experience for all fund management (for instance). 

Part of the reason the above “this is extra stuff” is not listed - as soon as the basics are in place (which is frustrating since even the code needs a LOT of cleanup), we need to see how to migrate to a proper, simple, UX on a per account basis. 

The reality is that as we go through and get the basics in place, we are actually seeing just how massive the complexity problem here is. (And that is even without corner cases). So it is exceptionally valuable from that POV since there it is easy talking without using, and when you actually use it (and other does), you spot where things are just not simple and clear. 

TL;DR staking UX, o, staking UX - I wish I could love you dearly.
